### ISSUE TITLE
Generics: False negative on type constraint validation involving a conditional type

### ISSUE THREAD

**Issue by UselessPickles** [microsoft/TypeScript#25413]
**TypeScript Version:**  2.9.2

**Search Terms:** 
conditional type constraint

**Code**

```ts
declare class Model<M extends MR, MR extends {}> {
    // Compiler error: 
    // Type 'M[K]' does not satisfy the constraint 'K extends keyof MR ? MR[K] : M[K]'.
    public getField2<K extends keyof M>(): Field<M[K], K extends keyof MR ? MR[K] : M[K]>
}

declare class Field<T extends TR, TR> {
}
```

**Expected behavior:**
No compiler error.

Because `M extends MR`, and `K extends keyof M`:
* If `K extends keyof MR`, then I expect `M[K]` to satisfy the constraint `MR[K]`.
* Else, I expect `M[K]` to satisfy the constraint `M[K]`.
* Therefore, it seems that `M[K]` should satisfy the constraint `K extends keyof MR ? MR[K] : M[K]`

**Actual behavior:**
Compiler error: Type 'M[K]' does not satisfy the constraint 'K extends keyof MR ? MR[K] : M[K]'.

**Playground Link:** 
https://www.typescriptlang.org/play/#src=declare%20class%20Model%3CM%20extends%20MR%2C%20MR%20extends%20%7B%7D%3E%20%7B%0D%0A%20%20%20%20%2F%2F%20Compiler%20error%3A%20%0D%0A%20%20%20%20%2F%2F%20Type%20'M%5BK%5D'%20does%20not%20satisfy%20the%20constraint%20'K%20extends%20keyof%20MR%20%3F%20MR%5BK%5D%20%3A%20M%5BK%5D'.%0D%0A%20%20%20%20public%20getField2%3CK%20extends%20keyof%20M%3E()%3A%20Field%3CM%5BK%5D%2C%20K%20extends%20keyof%20MR%20%3F%20MR%5BK%5D%20%3A%20M%5BK%5D%3E%0D%0A%7D%0D%0A%0D%0Adeclare%20class%20Field%3CT%20extends%20TR%2C%20TR%3E%20%7B%0D%0A%7D

**Related Issues:** 
Possibly?
* https://github.com/Microsoft/TypeScript/issues/24190
* https://github.com/Microsoft/TypeScript/issues/23843

**Comment by UselessPickles** [comment]
Interesting followup. As of now, at least, with TypeScript 3.3.3, I can get the correct behavior with a small change: use `MR extends Record<string, any>` instead of `MR extends {}` as my type constraint. The original `MR extends {}` seems to cause problems because type `{}` has no keys:

> Type 'M[K]' does not satisfy the constraint 'K extends keyof MR ? MR[K] : M[K]'.
>   Type 'MR[K]' is not assignable to type 'K extends keyof MR ? MR[K] : M[K]'.
>     Type '{}[K]' is not assignable to type 'K extends keyof MR ? MR[K] : M[K]'.

The change to `MR extends Record<string, any>` seems to clarify that `MR` is guaranteed to have string keys.

I'm still a bit unclear on when it is appropriate to use type `{}` as a constraint. Is there still a bug that needs to be investigated to get my original code to compile as I expect it to, or is the change to `MR extends Record<string, any>`actually the correct solution?

**Comment by jack-williams** [comment]
I think should be covered by #26933 provided you convert the conditional type to be non-distributive.

**Comment by UselessPickles** [comment]
@jack-williams I'm not familiar with what a "distributive" vs "non-distributive" type is. Could you elaborate on what you mean by "convert the conditional type to be non-distributive."?

**Comment by jack-williams** [comment]
@UselessPickles Here is a relevant [comment ](https://github.com/Microsoft/TypeScript/issues/30378#issuecomment-472627447)that should have the information (or at least point it).

**Comment by UselessPickles** [comment]
Thanks. I like the trick for preventing distribution (in [this comment](https://github.com/Microsoft/TypeScript/issues/29535#issuecomment-456672565), reference by the comment you pointed me to) :)

**Comment by weswigham** [comment]
@jack-williams is correct - #26933 + using a non-distributive conditional does fix this. The reason a distributive conditional can't be assumed to work is because the `K` in either branch of ` K extends keyof MR ? MR[K] : M[K]` is some specific subtype of `K` (thanks to distribution), while the `K` in `M[K]` is just the input `K`. So if, for example, your input `K` was `"ok" | "not ok"`, the `K` in each branch could be one of `"ok"` _or_ `"not ok"`, and the union of both is not assignable to each individually (it contains the other option, ofc).
